clients/src/test/java/org/apache/kafka/common/requests/BrokerRegistrationRequestTest.java

```diff
@@ -51,7 +60,84 @@ public void testBasicBuild(short version) {
             .setListeners(new BrokerRegistrationRequestData.ListenerCollection())
             .setRack("a")
             .setPreviousBrokerEpoch(1L);
-        BrokerRegistrationRequest.Builder builder = new BrokerRegistrationRequest.Builder(data);
+        BrokerRegistrationRequestData data2 = readSerializedRequest(version, data);
+
+        assertEquals(0, data2.brokerId(), "Unexpected broker ID in " + data2);
+        assertEquals("test", data2.clusterId(), "Unexpected cluster ID in " + data2);
+        assertEquals(incarnationId, data2.incarnationId(), "Unexpected incarnation ID in " + data2);
+        assertEquals("a", data2.rack(), "Unexpected rack in " + data2);
+        if (version >= 3) {
+            assertEquals(1, data2.previousBrokerEpoch(), "Unexpected previousBrokerEpoch in " + data2);
+        } else {
+            assertEquals(-1, data2.previousBrokerEpoch(), "Unexpected previousBrokerEpoch in " + data2);
+        }
+    }
+
+    @Test
+    public void testV0SerializationFailsWithZkMigrationEnabled() {
+        assertThrows(UnsupportedVersionException.class,
+            () -> readSerializedRequest((short) 0,
+                new BrokerRegistrationRequestData().
+                    setIsMigratingZkBroker(true)));
+    }
+
+    @ParameterizedTest
+    @MethodSource("BrokerRegistrationRequestVersionsWithoutV0")
+    public void testSuppressFeaturesWithMinVersion0OnOlderVersions(short version) {
```

Comment:
It's not clear what older versions are. How about testSuppressFeaturesWithMinVersion0BeforeV4 ?